src/Protocol.h

```diff
@@ -0,0 +1,80 @@
+﻿#pragma once
+
+#include <span>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+inline std::string readPrefixedString(std::span<const uint8_t>& buf)
+{
+    if (buf.empty())
+        throw std::runtime_error("Buffer too small for string length");
+
+    uint8_t len = buf[0];
+    if (len > 128)
+        throw std::runtime_error("String length too large");
+
+    if (buf.size() < 1 + len)
+        throw std::runtime_error("Buffer too small for string content");
+
+    std::string result(reinterpret_cast<const char*>(buf.data() + 1), len);
+    buf = buf.subspan(1 + len);
+    return result;
+}
+
+inline void writePrefixedString(std::vector<uint8_t>& buf, const std::string& str)
+{
+    if (str.size() > 128)
+        throw std::runtime_error("String length too large");
+
+    buf.push_back(static_cast<uint8_t>(str.size()));
+    buf.insert(buf.end(), str.begin(), str.end());
+}
+
+#define VARIABLE_SIZE_PAYLOAD_ONE_ARRAY_AT_END(maxSize) \
```

Comment:
Macros! Though fair enough - this is clean enough to also work. 
Another idea - how about declaring a `std::array<TEndArrayElement, maxSize>` for data - considering we know that max bounds and length in the impls I've seen thus far? Or implementing seperate `serialize` `deserialize` functions with the macro instead and leverage STL types like `std::string` and `std::vector` for storage?

Reply:
Using `std::array<TEndArrayElement, maxSize>` expands the array variable size to `maxSize * sizeof(TEndArrayElement)`. Using the empty array declaration now does not influence the struct size but still allows data from that offset to be accessed without casting `this` to `uint8_t*`.

As for serialize/deserialize, I think it is not needed for these kind of payloads with only one variable size array. Remember, I like optimizing stuff, so I'd like stuff done with least heap allocations/reallocations as possible and less binary instructions on Release if possible.

Reply:
I see. Guess using raw `new` or not going for the same seralize pattern triggered some OCD in me. 

I'd agree with you on the usage here - this seems like a very reasonable tradeoff.